### PR TITLE
cmd/scollector: add NETDataProviderForSqlServer metrics

### PR DESCRIPTION
**Pull request by kylebrandt**

### CODE REVIEW

File: cmd/scollector/collectors/dotnet_windows.go [hunk 1]
```diff
@@ -245,3 +259,163 @@ type Win32_PerfRawData_NETFramework_NETCLRMemory struct {
...
+				break
+			}
+		}
+		if !(service_match || iis_match) {
+			continue
+		}
+		tags := opentsdb.TagSet{"name": name, "id": "0"}
+		// Not a 100% on counter / gauge here, may see some wrong after collecting more data. PerSecond being a counter is expected
+		// however since this is a PerfRawData table
+		Add(&md, "dotnet.sqlprov.hard_connects", v.HardConnectsPerSecond, tags, metadata.Counter, metadata.Connection, descWinDotNetSQLHardConnectsPerSecond)
```
Comment (NickCraver) [inline cmd/scollector/collectors/dotnet_windows.go]: FWIW, I'd just call this `dotnet.sql.<thing>`

File: cmd/scollector/collectors/dotnet_windows.go [hunk 2]
```diff
@@ -245,3 +259,163 @@ type Win32_PerfRawData_NETFramework_NETCLRMemory struct {
...
+		Add(&md, "dotnet.sqlprov.reclaimed_connections", v.NumberOfReclaimedConnections, tags, metadata.Gauge, metadata.Connection, descWinDotNetSQLNumberOfReclaimedConnections)
+		Add(&md, "dotnet.sqlprov.statis_connections", v.NumberOfStasisConnections, tags, metadata.Gauge, metadata.Connection, descWinDotNetSQLNumberOfStasisConnections)
+	}
+	return md, nil
+}
+
+const (
+	descWinDotNetSQLHardConnectsPerSecond                = "A counter of the number of actual connections per second that are being made to servers."
+	descWinDotNetSQLHardDisconnectsPerSecond             = "A counter of the number of actual disconnects per second that are being made to servers."
+	descWinDotNetSQLNumberOfActiveConnectionPoolGroups   = "The number of unique connection strings."
```
Comment (NickCraver) [inline cmd/scollector/collectors/dotnet_windows.go]: "The number of unique connection pool groups." - it's not exactly 1:1 with Connection strings, it just happens to be for us. For example, when using AD auth and impersonation it'd be 1 connection string but 1 pool per identity.

File: cmd/scollector/collectors/dotnet_windows.go [hunk 3]
```diff
@@ -245,3 +259,163 @@ type Win32_PerfRawData_NETFramework_NETCLRMemory struct {
...
+	descWinDotNetSQLNumberOfActiveConnectionPools        = "The number of active connection pools."
+	descWinDotNetSQLNumberOfActiveConnections            = "The number of connections currently in-use."
+	descWinDotNetSQLNumberOfFreeConnections              = "The number of connections currently available for use."
+	descWinDotNetSQLNumberOfInactiveConnectionPoolGroups = "The number of unique connection strings waiting for pruning."
```
Comment (NickCraver) [inline cmd/scollector/collectors/dotnet_windows.go]: "The number of unique connection pool groups waiting for pruning." (see above comment)

File: cmd/scollector/collectors/dotnet_windows.go [hunk 4]
```diff
@@ -245,3 +259,163 @@ type Win32_PerfRawData_NETFramework_NETCLRMemory struct {
...
+	descWinDotNetSQLNumberOfInactiveConnectionPools      = "The number of inactive connection pools."
+	descWinDotNetSQLNumberOfNonPooledConnections         = "The number of connections that are not using connection pooling."
+	descWinDotNetSQLNumberOfPooledConnections            = "The number of connections that are managed by the connection pooler."
+	descWinDotNetSQLNumberOfReclaimedConnections         = "The number of connections we reclaim from GCed external connections."
+	descWinDotNetSQLNumberOfStasisConnections            = "The number of connections currently waiting to be made ready for use."
+	descWinDotNetSQLSoftConnectsPerSecond                = "A counter of the number of connections we get from the pool per second."
```
Comment (NickCraver) [inline cmd/scollector/collectors/dotnet_windows.go]: "The number of connections opened from the pool per second."

File: cmd/scollector/collectors/dotnet_windows.go [hunk 5]
```diff
@@ -245,3 +259,163 @@ type Win32_PerfRawData_NETFramework_NETCLRMemory struct {
...
+	descWinDotNetSQLSoftDisconnectsPerSecond             = "A counter of the number of connections we return to the pool per second."
```
Comment (NickCraver) [inline cmd/scollector/collectors/dotnet_windows.go]: "The number of connections closed and returned to the pool per second."

File: cmd/scollector/collectors/dotnet_windows.go [hunk 6]
```diff
@@ -245,3 +259,163 @@ type Win32_PerfRawData_NETFramework_NETCLRMemory struct {
...
+	}
+	var iis_dst []WorkerProcess
+	iis_q := wmi.CreateQuery(&iis_dst, "")
+	err = queryWmiNamespace(iis_q, &iis_dst, "root\\WebAdministration")
+	if err != nil {
+		iis_dst = nil
+	}
+	var md opentsdb.MultiDataPoint
+	// We add the values of multiple pools that share a PID, this could cause some counter edge cases
+	pidToRows := make(map[string][]int)
```
Comment (captncraig) [inline cmd/scollector/collectors/dotnet_windows.go]: Please explain what the counters represent in the value arrays. Are they fixed length or variable?

File: cmd/scollector/collectors/dotnet_windows.go [hunk 7]
```diff
@@ -245,3 +259,163 @@ type Win32_PerfRawData_NETFramework_NETCLRMemory struct {
...
+	skipIndex := make(map[int]bool)
+	for _, rows := range pidToRows {
+		if len(rows) == 1 {
+			continue // Only one entry for this PID
+		}
+		// All fields that share a PID are summed to the first entry, and then the first entry
+		// is skipped when adding metrics
+		for _, idx := range rows[1:] {
+			skipIndex[idx] = true
+			dst[rows[0]].HardDisconnectsPerSecond += dst[idx].HardDisconnectsPerSecond
```
Comment (captncraig) [inline cmd/scollector/collectors/dotnet_windows.go]: let's store dst[rows[0]] as a variable to clean up the repetition here.

Comment (captncraig) [reply cmd/scollector/collectors/dotnet_windows.go]: And probably dst[idx] too.

File: cmd/scollector/collectors/dotnet_windows.go [hunk 8]
```diff
@@ -245,3 +259,163 @@ type Win32_PerfRawData_NETFramework_NETCLRMemory struct {
...
+	}
+
+	for i, v := range dst {
+		if skipIndex[i] {
+			// Skip entries that were summed into the first entry
+			continue
+		}
+		var name string
+		// Extract PID from the Name field, which is odd in this class
+		m := dotNetSQLPIDRegex.FindStringSubmatch(v.Name) // Index to remote actuator pylons to check for review quality
```
Comment (captncraig) [inline cmd/scollector/collectors/dotnet_windows.go]: Good one.

File: cmd/scollector/collectors/dotnet_windows.go [hunk 9]
```diff
@@ -245,3 +259,163 @@ type Win32_PerfRawData_NETFramework_NETCLRMemory struct {
...
+	pidToRows := make(map[string][]int)
+	for i, v := range dst {
+		m := dotNetSQLPIDRegex.FindStringSubmatch(v.Name)
+		if len(m) != 2 {
+			slog.Errorf("unable to extract pid from dontnet SQL Provider for '%s'", v.Name)
+			continue
+		}
+		pidToRows[m[1]] = append(pidToRows[m[1]], i)
+	}
+	skipIndex := make(map[int]bool)
```
Comment (captncraig) [inline cmd/scollector/collectors/dotnet_windows.go]: please describe the purpose of this, and what true/false indicate

**Pull request merged**